Objects/obmalloc.c

```diff
@@ -717,6 +717,70 @@ PyObject_Free(void *ptr)
 #  define LIKELY(value) (value)
 #endif
 
+
+#if WITH_FREELISTS
+void *
+_PyFreeList_HalfFillAndAllocate(_PyFreeList *list)
+{
+    assert(list->ptr == NULL);
+    if (list->capacity < 4) {
+        return PyObject_Malloc(list->size);
+    }
+    uint32_t i = 0;
+    for (; i < list->space>>1; i++) {
+        void* ptr = PyObject_Malloc(list->size);
+        if (ptr == NULL) {
+            break;
+        }
+        *((void**)ptr) = list->ptr;
+        list->ptr = ptr;
+    }
+    if (i == 0) {
+        return NULL;
+    }
+    list->space -= (i-1);
```

Comment:
Why `i-1`?

Reply:
Yeah it looks wrong. I don't understand this part too:

```
        *((void**)ptr) = list->ptr;
        list->ptr = ptr;          
```

Reply:
oh maybe the -1 is to account for the result being allocated in the next line.